### PR TITLE
gnrc_rpl: introduce control plane traffic statistics

### PR DESCRIPTION
**Pull request by cgundogan**
Provide statistics about RX/TX DIOs, DISs, DAOs, DAO-ACKs.

~~based on #5178~~

### CODE REVIEW

**Comment by BytesGalore** [comment]
I guess its not waiting anymore.

File: sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c [hunk 1]
```diff
@@ -278,22 +292,31 @@ void gnrc_rpl_recv_DIS(gnrc_rpl_dis_t *dis, kernel_pid_t iface, ipv6_addr_t *src
...
         for (uint8_t i = 0; i < GNRC_RPL_INSTANCES_NUMOF; ++i) {
-            if ((gnrc_rpl_instances[i].state != 0)
+            if (gnrc_rpl_instances[i].state != 0) {
+#ifdef MODULE_NETSTATS_RPL
+                gnrc_rpl_instances[i].stats.dis_rx_count++;
+                gnrc_rpl_instances[i].stats.dis_rx_bytes += len;
+#endif
                 /* a leaf node should only react to unicast DIS */
-                 && (gnrc_rpl_instances[i].dodag.node_status != GNRC_RPL_LEAF_NODE)) {
+                if(gnrc_rpl_instances[i].dodag.node_status != GNRC_RPL_LEAF_NODE) {
```
Comment (BytesGalore) [inline sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c]: please add a whitespace

**Comment by BytesGalore** [comment]
Don't know if it makes sense, but how about also counting errors, e.g. invalid DIOs/DAOs?

**Comment by cgundogan** [comment]
@BytesGalore rebased onto master and did a little overhaul. The statistics aren't bound to the instances anymore, because they would get lost otherwise when a node leaves the dodag.

You idea about counting errors is also nice, but let's concentrate on this for another PR (: getting to big again.

You can test by compiling `gnrc_networkin` and calling `rpl stats` from any node.

File: sys/include/net/netstats.h [hunk 2]
```diff
@@ -52,6 +53,45 @@ typedef struct {
...
+    /* DAO-ACK */
+    uint32_t dao_ack_rx_ucast_count;    /**< unicast dao_ack received in packets */
+    uint32_t dao_ack_rx_ucast_bytes;    /**< unicast dao_ack received in bytes */
+    uint32_t dao_ack_rx_mcast_count;    /**< multicast dao_ack received in packets */
+    uint32_t dao_ack_rx_mcast_bytes;    /**< multicast dao_ack received in bytes */
+    uint32_t dao_ack_tx_ucast_count;    /**< unicast dao_ack sent in packets */
+    uint32_t dao_ack_tx_ucast_bytes;    /**< unicast dao_ack sent in bytes */
+    uint32_t dao_ack_tx_mcast_count;    /**< multicast dao_ack sent in packets */
+    uint32_t dao_ack_tx_mcast_bytes;    /**< multicast dao_ack sent in bytes*/
+} netstats_rpl_t;
```
Comment (OlegHahm) [inline sys/include/net/netstats.h]: Wouldn't it make maybe more sense to put this into some RPL specific header?

Comment (BytesGalore) [inline sys/include/net/netstats.h]: +1

Comment (BytesGalore) [inline sys/include/net/netstats.h]: would it make sense to introduce a counter type to easy switching smaller/larger nummeric types, e.g. `uint16_t`, like:
``` C
typedef uint16_t netstat_rpl_cnt_t;
typedef struct {
    /* DIO */
    netstat_rpl_cnt_t dio_rx_ucast_count;        /**< unicast dio received in packets */
...
}netstats_rpl_t;
```
?

File: sys/net/gnrc/routing/rpl/gnrc_rpl.c [hunk 3]
```diff
@@ -149,17 +157,17 @@ static void _receive(gnrc_pktsnip_t *icmpv6)
         case GNRC_RPL_ICMPV6_CODE_DIO:
             DEBUG("RPL: DIO received\n");
             gnrc_rpl_recv_DIO((gnrc_rpl_dio_t *)(icmpv6_hdr + 1), iface, &ipv6_hdr->src,
```
Comment (BytesGalore) [inline sys/net/gnrc/routing/rpl/gnrc_rpl.c]: I wonder if we could just pass the `&ipv6_hdr` to the functions and access `->src`, `->dst` and `->len` inside the function.

**Comment by kYc0o** [comment]
@cgundogan do you think you can address the comments during the next week?

**Comment by kYc0o** [comment]
Well, let's see for the next release.

File: sys/net/gnrc/routing/rpl/gnrc_rpl.c (hunk 3 above)
Comment (cgundogan) [inline sys/net/gnrc/routing/rpl/gnrc_rpl.c]: Yeah that would definitely make sense. Will adjust in another PR.

**Comment by cgundogan** [comment]
I moved the rpl related netstats struct into a separate file as suggested by @OlegHahm

**Comment by cgundogan** [comment]
Somehow the comment of @BytesGalore regarding a configurable struct member type got lost (commented on commit? (; ). But we can also do this as a follow-up, what do you think @BytesGalore ?

**Comment by BytesGalore** [comment]
@cgundogan you mean https://github.com/RIOT-OS/RIOT/pull/5381/commits/8f1d4380f621e793e49a892d978d9b4b4a71ac17#r65735118 :
``` C
typedef uint16_t netstat_rpl_cnt_t;
typedef struct {
    /* DIO */
    netstat_rpl_cnt_t dio_rx_ucast_count;        /**< unicast dio received in packets */
...
}netstats_rpl_t;
```
no objection to put/discuss it in a followup PR

**Comment by BytesGalore** [comment]
I would say please squash then ACK from my side

**Comment by OlegHahm** [comment]
@cgundogan, ping!

**Comment by BytesGalore** [comment]
@cgundogan I still would say please squash then ACK from my side

**Comment by cgundogan** [comment]
@BytesGalore squashed

**Pull request merged**